modules/indexing/src/test/java/org/apache/ignite/cache/query/RepeatedFieldIndexQueryTest.java

```diff
@@ -0,0 +1,337 @@
+
+    /** */
+    private static final int CNT = 10_000;
+
+    /** */
+    private static IgniteCache<Integer, Person> cache;
+
+    /** */
+    @Parameterized.Parameter
+    public String idxName;
+
+    /** */
+    @Parameterized.Parameter(1)
+    public String fldName;
+
+    /** */
+    @Parameterized.Parameters(name = "idx={0} fldName={1}")
+    public static List<Object[]> params() {
+        return F.asList(
+            new Object[] {ID_IDX, "id"},
+            new Object[] {DESC_ID_IDX, "descId"}
+        );
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        Ignite crd = startGrids(2);
+
+        cache = crd.cache(CACHE);
+
+        for (int i = 0; i < CNT; i++)
+            cache.put(i, new Person(i));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        CacheConfiguration<?, ?> ccfg1 = new CacheConfiguration<>()
+            .setName(CACHE)
+            .setIndexedTypes(Integer.class, Person.class);
+
+        cfg.setCacheConfiguration(ccfg1);
+
+        return cfg;
+    }
+
+    /** */
+    @Test
+    public void testRangeQueriesWithTwoCriteriaSingleField() {
+        int lower = new Random().nextInt(CNT / 2);
+        int upper = CNT / 2 + new Random().nextInt(CNT / 2 - 1);
+
+        List<T2<RangeIndexQueryCriterion, RangeIndexQueryCriterion>> cc = new ArrayList<>();
+
+        Stream.of(new T2<>(lower, upper), new T2<>(upper, lower)).forEach(pair -> {
+            for (int i = 0; i < 6; i++) {
+                for (int j = 0; j < 6; j++) {
+                    RangeIndexQueryCriterion c1 = (RangeIndexQueryCriterion)criterion(i, fldName, pair.get1(), pair.get2());
+                    RangeIndexQueryCriterion c2 = (RangeIndexQueryCriterion)criterion(j, fldName, pair.get2(), pair.get1());
+
+                    cc.add(new T2<>(c1, c2));
+                }
+            }
+        });
+
+        cc.forEach(c -> checkTwoCriteria(c.get1(), c.get2()));
+    }
+
+    /** */
+    @Test
+    public void testMultipleEqualsCriteria() {
+        int lower = new Random().nextInt(CNT / 2);
+        int upper = CNT / 2 + new Random().nextInt(CNT / 2 - 1);
+
+        checkEqualsCriteria(lower, upper, 0, CNT);
+        checkEqualsCriteria(lower, upper, CNT, 0);
+        checkEqualsCriteria(upper, lower, 0, CNT);
+        checkEqualsCriteria(upper, lower, CNT, 0);
+    }
+
+    /** */
+    private void checkEqualsCriteria(int eq1, int eq2, int from, int to) {
+        IndexQuery<Integer, Person> qry = new IndexQuery<Integer, Person>(Person.class, idxName)
+            .setCriteria(eq(fldName, eq1), eq(fldName, eq2), between(fldName, from, to));
+
+        assertTrue(qry.getCriteria().toString(), cache.query(qry).getAll().isEmpty());
+    }
+
+    /** */
+    @Test
+    public void testCommonBoundary() {
```

Comment:
test check nothing common here %)

Reply:
Why? upper == lower in this test.

Reply:
Renamed `lower` and `upper` to `boundary`.